[view-transitions] Clear view transition on navigation

nt1m opened the pull request:
#### f6ed29f91fc8f5d7fbde9c3d4a825b07dd6111fb
<pre>
[view-transitions] Clear view transition on navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=272792">https://bugs.webkit.org/show_bug.cgi?id=272792</a>
<a href="https://rdar.apple.com/126590072">rdar://126590072</a>

Reviewed by Chris Dumez.

The VT crashtests were causing the following test to flakily crash. To fix this, clear the view transition when the user navigates by
using ActiveDOMObject and defining a stop method.

Also protect the document in `ViewTransition::callUpdateCallback`.

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::clearViewTransition):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::stop):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::document const):
(WebCore::ViewTransition::protectedDocument const):
* Source/WebCore/dom/ViewTransition.idl:

Canonical link: <a href="https://commits.webkit.org/277958@main">https://commits.webkit.org/277958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/334edf997416e14f4865cc5e56ff05ea0c282081

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51769 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45121 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40085 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25902 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21194 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43462 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7169 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45288 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43967 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53678 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47401 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42495 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46371 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10793 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26181 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -60,7 +60,8 @@ static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(
 ViewTransition::ViewTransition(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
-    : m_document(document)
+    : ActiveDOMObject(document)
+    , m_document(document)
```
Shouldn't be needed, you can get your document from ActiveDOMObject.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -72,7 +73,9 @@ ViewTransition::~ViewTransition() = default;
 Ref<ViewTransition> ViewTransition::create(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
 {
-    return adoptRef(*new ViewTransition(document, WTFMove(updateCallback)));
+    auto viewTransition = adoptRef(*new ViewTransition(document, WTFMove(updateCallback)));
```
`auto` -> `Ref`

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -384,8 +386,8 @@ ExceptionOr<void> ViewTransition::captureOldState()
         capture.oldOverflowRect = captureOverflowRect(element);
-        if (m_document->frame())
-            capture.oldImage = snapshotElementVisualOverflowClippedToViewport(*m_document->frame(), element, capture.oldOverflowRect);
+        if (document()->frame())
```
```
if (RefPtr frame = document()->frame()))
    capture.oldImage = snapshotElementVisualOverflowClippedToViewport(*frame, element, capture.oldOverflowRect);
```

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -547,10 +548,10 @@ void ViewTransition::activateViewTransition()
         return;
 
-    RefPtr documentElement = m_document->documentElement();
+    RefPtr documentElement = protectedDocument()->documentElement();
```
We don't need `protectedDocument()` here, `documentElement()` is inline and trivial.

cdumez commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -151,10 +152,13 @@ class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<
 
     RenderViewTransitionCapture* viewTransitionNewPseudoForCapturedElement(Element&);
 
+    void stop() final;
```
Should likely be private.

cdumez reviewed (approved):
LGTM with nits.

nt1m committed:
[view-transitions] Clear view transition on navigation
https://bugs.webkit.org/show_bug.cgi?id=272792
rdar://126590072

Reviewed by Chris Dumez.

The VT crashtests were causing the following test to flakily crash. To fix this, clear the view transition when the user navigates by
using ActiveDOMObject and defining a stop method.

Also protect the document in `ViewTransition::callUpdateCallback`.

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::clearViewTransition):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::stop):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::document const):
(WebCore::ViewTransition::protectedDocument const):
* Source/WebCore/dom/ViewTransition.idl:

Canonical link: https://commits.webkit.org/277958@main
